src/ast/passes/resource_analyser.cpp

```diff
@@ -29,16 +30,22 @@ std::string get_literal_string(Expression &expr)
 
 } // namespace
 
-ResourceAnalyser::ResourceAnalyser(Node *root) : root_(root), probe_(nullptr)
+ResourceAnalyser::ResourceAnalyser(Node *root, std::ostream &out)
+    : root_(root), out_(out), probe_(nullptr)
 {
 }
 
-RequiredResources ResourceAnalyser::analyse()
+std::pair<RequiredResources, int> ResourceAnalyser::analyse()
 {
   Visit(*root_);
   prepare_seq_printf_ids();
 
-  return std::move(resources_);
+  if (!err_.str().empty())
+  {
+    out_ << err_.str();
+  }
+
+  return { std::move(resources_), !err_.str().empty() };
```

Comment:
Since this is returning `bool`, shouldn't the return type be `std::pair<RequiredResources, bool>`? Or, maybe even better, `std::optional<RequiredResources>`?

Reply:
All other analysers (`FieldAnalyser`, `PortabilityAnalyser`, `SemanticAnalyser`) return `int`. In case of `SemanticAnalyser` it's because it returns the number of the pass it failed in, the other ones I assume are `int` was chosen to easily allow multiple error codes in the future if necessary. Either way, wouldn't it be better to keep it consistent here?

Reply:
Well, consistency is nice but we don't really make use of the error codes anyway. And since `ResourceAnalyser` already returns something different, I wouldn't try too hard here to be similar to other analysers.

Reply:
OK, that makes sense.

Reply:
Also the passes are handled by the pass manager nowdays and that has its own error handling (even though its a bit shitty too). So what you do here doesn't matter too much.